### PR TITLE
Add inject support to namespaces

### LINKED ISSUE #3255: auto-proxy support for namespaces
### What problem are you trying to solve?

Now that auto-proxy works, please teach `linkerd inject` how to annotate a namespace like:

`kubectl get ns/default -o yaml | linkerd inject -`

So that anything added to said namespace magically gets linkerd mesh support!

### Any alternatives you've considered?

Currently, we are running a `kubectl patch`, but this is not as clean.

This is as of linkerd v2.4 stable.

### PR DESCRIPTION
**Pull request by mayankshah1607**
Fixes #3255 

This PR adds support for `linkerd inject` to annotate namespaces as follows : 
`kubectl get ns/default -o yaml | linkerd inject -`
Anything added to the said namespace gets linkerd support.

- Added method `AppendNsAnnotation()` for `ResourceConfig` objects
- Added method `IsNamespace()` for `ResourceConfig` objects - checks if a given config is of Kind namespace
- Updated `linkerd install` golden files
- Updated `inject.go` to make namespace annotations come in through patching - https://github.com/Pothulapati/linkerd2/commit/b49fcdc096c9fd70e959d2ab34f7618155c8e10b
- Relevant unit tests have been added with `overrideAnnotations` as well

### CODE REVIEW

File: pkg/inject/inject.go [hunk 1]
```diff
@@ -110,6 +101,18 @@ type ResourceConfig struct {
 	}
 }
 
+// Workload contains the information for a given conf
+type Workload struct {
```
Comment (Pothulapati) [inline pkg/inject/inject.go]: As we can check the `kind` directly by 
```go
//IsWorkloadNamespace checks if a given config is of type namespace
func (conf *ResourceConfig) IsWorkloadNamespace() bool {
	return strings.ToLower(conf.workload.metaType.Kind) == k8s.Namespace
}
```

I don't see a point in making `workload` public and taking all this hassle.

File: cli/cmd/inject.go [hunk 2]
```diff
@@ -165,7 +171,9 @@ func (rt resourceTransformerInject) transform(bytes []byte) ([]byte, []inject.Re
 	}
 	log.Infof("patch generated for: %s", report.ResName())
 	log.Debugf("patch: %s", patchJSON)
+
```
Comment (Pothulapati) [inline cli/cmd/inject.go]: can you remove these unnecessary white spaces?

File: pkg/inject/inject.go [hunk 3]
```diff
@@ -869,3 +920,8 @@ func sortedKeys(m map[string]string) []string {
 
 	return keys
 }
+
+//IsNamespace checks if a given config is of type namespace
+func (conf *ResourceConfig) IsNamespace() bool {
```
Comment (Pothulapati) [inline pkg/inject/inject.go]: I dont think we need this anymore. We can use the `IsWorkloadNamespace()` everywhere as both of these functions are doing the same.

**Comment by Pothulapati** [comment]
This Looks good :100: 

@ihcsim can you take a look?

**Comment by Pothulapati** [comment]
Hmm, @ihcsim that would make the logic simple, b
ut a command like `linkerd inject --proxy-Image XYZ namespace.yaml` would not add the config annoations. 
This may create confusion now that the config annotations are honoured by the proxy-injector during injection.

**Comment by mayankshah1607** [comment]
I'll work on these changes and open another PR.

**Comment by cpretzer** [comment]
@mayankshah1607 do you mind pushing your commits so that we can take a look at the code where the namespace is `nil`?

**Comment by mayankshah1607** [comment]
Hi @cpretzer . I have pushed my commits - https://github.com/mayankshah1607/linkerd2/commit/24f0c0cae76ca159fead9be32c127bf90cea3ca3

**Comment by mayankshah1607** [comment]
Fixed it. The problem was that `conf.workload.obj` itself was `nil` because `conf.ParseMetaYAML()` was not configured to return a namespace kind ResourceConfig. 
Everything works as expected now.

Considering all the factors discussed above, I have opened a new PR as this one has too many unnecessary file changes dealing with the `GetPatch()` method. Hence, I shall be closing this one.

Please review my new PR - https://github.com/linkerd/linkerd2/pull/3607

**Pull request closed without merging**